### PR TITLE
usb: it82xx2: fix issues about the extended endpoint access and read func behavior

### PR DESCRIPTION
**Pull request by Chenhongren**
This PR fixs some issues including
- fix some issues related to extended endpoint enablement
- lock irq when accessing fifo ctrl register and ready bit
- return the available data length in the ep read function

Tested with: Google spikyrock project

### CODE REVIEW

**Review by tmon-nordic — DISMISSED** [review]

**Comment by Chenhongren** [comment]
Upload a slight change: return the correct type for the `get_fifo_ctrl()` function.

**Review by tmon-nordic — DISMISSED** [review]

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 1]
```diff
@@ -695,45 +676,151 @@ static inline void it82xx2_handler_out(uint8_t fifo_idx)
...
+	if (((ep_idx - 4) / 3 == 0)) {
+		enable = !!(epn_ext_ctrl1->fields.epn0_enable_bit);
+	} else if (((ep_idx - 4) / 3 == 1)) {
+		enable = !!(epn_ext_ctrl1->fields.epn3_enable_bit);
+	} else if (((ep_idx - 4) / 3 == 2)) {
+		enable = !!(epn_ext_ctrl1->fields.epn6_enable_bit);
+	} else {
+		enable = !!(epn_ext_ctrl1->fields.epn9_enable_bit);
+	}
+	return enable;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_it82xx2.c:725]:
```suggestion
	}

	return enable;
```

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 2]
```diff
@@ -695,45 +676,151 @@ static inline void it82xx2_handler_out(uint8_t fifo_idx)
...
+}
+
+static bool get_extend_ready_bit(const uint8_t ep_idx)
+{
+	struct usb_it82xx2_regs *const usb_regs = (struct usb_it82xx2_regs *)it82xx2_get_usb_regs();
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_it82xx2.c:730]: The return type is `struct usb_it82xx2_regs *`. Why is there a explicit conversion? Please clean it up overall in the file.

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 3]
```diff
@@ -695,45 +676,151 @@ static inline void it82xx2_handler_out(uint8_t fifo_idx)
...
+	struct epn_ext_ctrl_regs *ext_ctrl =
+		usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
+	int idx = ((ep_idx - 4) % 3) + 1;
+
+	return !!(ext_ctrl[idx].epn_ext_ctrl2 & BIT((ep_idx - 4) / 3));
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_it82xx2.c:735]: Conversion to bool type is described in e.g. C99 spec. Why is there `!!`?

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_it82xx2.c (hunk 3 above)
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c]: Fixed. Thanks.

File: drivers/usb/device/usb_dc_it82xx2.c (hunk 2 above)
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c]: removed.

File: drivers/usb/device/usb_dc_it82xx2.c (hunk 1 above)
Comment (Chenhongren) [reply drivers/usb/device/usb_dc_it82xx2.c]: done, thanks.

**Review by tmon-nordic — APPROVED** [review]

File: drivers/usb/device/usb_dc_it82xx2.c [hunk 4]
```diff
@@ -687,44 +668,150 @@ static inline void it82xx2_handler_out(uint8_t fifo_idx)
...
+}
+
+static uint16_t get_fifo_ctrl(const uint8_t fifo_idx)
+{
+	struct usb_it82xx2_regs *const usb_regs = it82xx2_get_usb_regs();
+	volatile uint8_t *ep_fifo_ctrl = usb_regs->fifo_regs[EP_EXT_REGS_BX].fifo_ctrl.ep_fifo_ctrl;
+	uint8_t fifon_ctrl = (fifo_idx - 1) * 2;
+
+	if (fifo_idx == 0) {
+		LOG_ERR("Invalid fifo_idx 0x%x", fifo_idx);
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_it82xx2.c:738]: __ASSERT?

**Review by jfischer-no — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**